clippy_lints/src/redundant_type_annotations.rs

```diff
@@ -0,0 +1,143 @@
+    /// ```
+    #[clippy::version = "1.70.0"]
+    pub REDUNDANT_TYPE_ANNOTATIONS,
+    pedantic,
+    "Warns about needless / redundant type annotations."
+}
+declare_lint_pass!(RedundantTypeAnnotations => [REDUNDANT_TYPE_ANNOTATIONS]);
+
+fn is_redundant_in_resolved_path<'tcx>(cx: &LateContext<'tcx>, res: hir::def::Res, func_return_type: Ty<'tcx>) -> bool {
+    // type annotation is primitive
+    if_chain! {
+        if let hir::def::Res::PrimTy(primty) = res;
+
+        if func_return_type.is_primitive();
+        if let Some(func_return_type_sym) = func_return_type.primitive_symbol();
+
+        then {
+            return primty.name() == func_return_type_sym;
+        }
+    }
+
+    // type annotation is any other non generic type
+    if_chain! {
+        if let hir::def::Res::Def(_, defid) = res;
+        if let Some(annotation_ty) = cx.tcx.type_of(defid).no_bound_vars();
+
+        then {
+            return annotation_ty == func_return_type
+        }
+    }
+
+    false
+}
+
+fn is_redundant_in_func_call<'tcx>(
+    cx: &LateContext<'tcx>,
+    res: hir::def::Res,
+    func_return_path: &hir::QPath<'tcx>,
+) -> bool {
+    // TODO: Problem with something like
+    // let a: String = StructTest::func() where func returns String::from
+    // The problem is that the DefId that I get refers to the struct itself and not to string
+
+    match func_return_path {
+        // let a: String = f(); where f: fn f() -> String
+        hir::QPath::Resolved(_, resolved_path) => {
+            if_chain! {
+                if let hir::def::Res::Def(_, defid) = resolved_path.res;
+                if let Some(middle_ty_init) = cx.tcx.type_of(defid).no_bound_vars();
+                if middle_ty_init.is_fn();
+                if let Some(init_return_type) = middle_ty_init.fn_sig(cx.tcx).output().no_bound_vars();
+                then {
+                    return is_redundant_in_resolved_path(cx, res, init_return_type);
+                }
+            }
+
+            false
+        },
+        // let a: String = String::new();
+        hir::QPath::TypeRelative(func_hir_ty, _) => {
+            if_chain! {
+                if let hir::def::Res::Def(_, defid) = res;
+                if let Some(annotation_ty) = cx.tcx.type_of(defid).no_bound_vars();
+
+                if let hir::TyKind::Path(init_ty_path) = &func_hir_ty.kind;
+                if let hir::QPath::Resolved(_, resolved_init_ty_path) = init_ty_path;
+                if let hir::def::Res::Def(_, init_defid) = resolved_init_ty_path.res;
+                if let Some(init_ty) = cx.tcx.type_of(init_defid).no_bound_vars();
+
+                then {
+                    return annotation_ty == init_ty
+                }
+            }
+
+            false
+        },
+        hir::QPath::LangItem(..) => false,
+    }
+}
+
+impl LateLintPass<'_> for RedundantTypeAnnotations {
+    fn check_local<'tcx>(&mut self, cx: &LateContext<'tcx>, local: &'tcx rustc_hir::Local<'_>) {
+        if_chain! {
```

Comment:
Here we should also include a check if the code comes from a macro. I believe it should be:

```rs
local.span.from_expansion()
```

or something close to that :)

Reply:
Done, thanks